apis/python/src/tiledbsoma/_index_util.py

```diff
@@ -3,26 +3,34 @@
 SOMATileDBContext which would otherwise ensue.
 """
 
+from typing import Optional
+
 import numpy as np
 import pandas as pd
+from somacore.query.types import IndexLike
 
 from tiledbsoma import pytiledbsoma as clib
 
 from .options import SOMATileDBContext
 
 
-def build_index(
-    keys: np.typing.NDArray[np.int64], context: SOMATileDBContext
-) -> clib.IntIndexer:
+def tiledbsoma_build_index(
+    keys: np.typing.NDArray[np.int64],
+    *,
+    context: Optional[SOMATileDBContext] = None,
+    thread_count: int = 4,
+) -> IndexLike:
     """Builds an indexer object compatible with :meth:`pd.Index.get_indexer`."""
     if len(np.unique(keys)) != len(keys):
```

Comment:
this test can (is) quite expensive for large indices -- np.unique effectively has to build a data structure similar to the index in order to perform the check. For example, on my x66 box, it takes 2-3 seconds to perform this test for 100M values.

Can we just use the index, which we are building, to do the same check?

E.g., if the index is a hash table insertion, where dups overwrite, then:
1. create the index
2. check that the resulting length == the input length

This would remove the duplication of work, which works against the entire point of having a fast indexer ;-)

Reply:
That sounds like a great idea @bkmartinjr. Let me see how quick I can implement that.